Spawning a child actor

Is there an example spawning a child actor somewhere?  I seem to be running into trouble with it, I am using the `ctx.ctx.spawn` facility:

```
[ERROR] [12/20/2016 14:32:09.070] [pingSystem-akka.actor.default-dispatcher-4] [akka://pingSystem/user/pingActor/$a] Wrap.message called with CheckZipStart(SessionId(272c5c45-9269-434c-a94c-893bc91bb815),94110,1482273129041,Actor[akka://pingSystem/system/testActor-1#496883580]); maybe the persistent typed behavior was deployed as a regular actor?
java.lang.IllegalStateException: Wrap.message called with CheckZipStart(SessionId(272c5c45-9269-434c-a94c-893bc91bb815),94110,1482273129041,Actor[akka://pingSystem/system/testActor-1#496883580]); maybe the persistent typed behavior was deployed as a regular actor?
	at com.nokia.ntp.ct.persistence.package$.impossible(package.scala:96)
	at com.nokia.ntp.ct.persistence.PersistentBehavior$Wrap.message(persistentBehavior.scala:110)
	at akka.typed.adapter.ActorAdapter$$anonfun$receive$1.applyOrElse(ActorAdapter.scala:32)
	at akka.actor.Actor$class.aroundReceive(Actor.scala:484)
	at akka.typed.adapter.ActorAdapter.aroundReceive(ActorAdapter.scala:9)
	at akka.actor.ActorCell.receiveMessage(ActorCell.scala:526)
	at akka.actor.ActorCell.invoke(ActorCell.scala:495)
	at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:257)
	at akka.dispatch.Mailbox.run(Mailbox.scala:224)
	at akka.dispatch.Mailbox.exec(Mailbox.scala:234)
	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```

--- a/src/main/java/com/nokia/ntp/ct/persistence/ActorContextAccess.java
+++ b/src/main/java/com/nokia/ntp/ct/persistence/ActorContextAccess.java
@@ -1,0 +1,48 @@
+/*
+ * Copyright 2016 Nokia Solutions and Networks Oy
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.nokia.ntp.ct.persistence;
+
+import akka.typed.adapter.ActorContextAdapter;
+
+/**
+ * The whole reason for the existence of this Java
+ * class is that `akka.typed.adapter.ActorContextAdapter`
+ * is package private, but we must access its underlying
+ * untyped ActorContext.
+ */
+final class ActorContextAccess {
+
+	public static akka.actor.ActorContext getUntypedActorContext(akka.typed.ActorContext<?> typedContext) {
+		if (typedContext instanceof ActorContextAdapter<?>) {
+			ActorContextAdapter<?> adapter = (ActorContextAdapter<?>) typedContext;
+			akka.actor.ActorContext result = adapter.akka$typed$adapter$ActorContextAdapter$$ctx;
+			if (result == null) {
+				throw new IllegalArgumentException("underlying ActorContext is null");
+			} else {
+				return result;
+			}
+		} else {
+			throw new UnsupportedOperationException(
+				"only adapted untyped ActorContexts permissible (got " +
+				typedContext +
+				" of class " +
+				typedContext.getClass() +
+				")"
+			);
+		}
+	}
+}

--- a/src/main/scala/com/nokia/ntp/ct/persistence/ActorDeployer.scala
+++ b/src/main/scala/com/nokia/ntp/ct/persistence/ActorDeployer.scala
@@ -1,0 +1,78 @@
+/*
+ * Copyright 2016 Nokia Solutions and Networks Oy
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.nokia.ntp.ct
+package persistence
+
+import akka.{ actor => au, typed => at }
+
+import shapeless.tag.@@
+
+/**
+ * Type class for abstracting over
+ * things which can start actors
+ * (e.g., ActorSystems or ActorContexts).
+ */
+trait ActorDeployer[D] {
+
+  def deploy[A](d: D)(props: au.Props @@ A, name: String): at.ActorRef[A] =
+    actorRef[A](deployUntyped(d)(props, name))
+
+  def deployAnonymous[A](d: D)(props: au.Props @@ A): at.ActorRef[A] =
+    actorRef[A](deployUntypedAnonymous(d)(props))
+
+  protected def deployUntyped(d: D)(props: au.Props, name: String): au.ActorRef
+
+  protected def deployUntypedAnonymous(d: D)(props: au.Props): au.ActorRef
+}
+
+object ActorDeployer {
+
+  def apply[A](implicit ev: ActorDeployer[A]): ActorDeployer[A] = ev
+
+  implicit val untypedActorSystemDeployerInstance: ActorDeployer[au.ActorSystem] = new ActorDeployer[au.ActorSystem] {
+    override def deployUntyped(d: au.ActorSystem)(props: au.Props, name: String): au.ActorRef =
+      d.actorOf(props, name)
+    override def deployUntypedAnonymous(d: au.ActorSystem)(props: au.Props): au.ActorRef =
+      d.actorOf(props)
+  }
+
+  implicit val untypedActorContextDeployer: ActorDeployer[au.ActorContext] = new ActorDeployer[au.ActorContext] {
+    override def deployUntyped(d: au.ActorContext)(props: au.Props, name: String): au.ActorRef =
+      d.actorOf(props, name)
+    override def deployUntypedAnonymous(d: au.ActorContext)(props: au.Props): au.ActorRef =
+      d.actorOf(props)
+  }
+
+  implicit def typedActorContextDeployerInstance[A]: ActorDeployer[at.ActorContext[A]] = new ActorDeployer[at.ActorContext[A]] {
+
+    override def deployUntyped(d: at.ActorContext[A])(props: au.Props, name: String): au.ActorRef =
+      untypedActorContextDeployer.deployUntyped(toUntyped(d))(props, name)
+
+    override def deployUntypedAnonymous(d: at.ActorContext[A])(props: au.Props): au.ActorRef =
+      untypedActorContextDeployer.deployUntypedAnonymous(toUntyped(d))(props)
+
+    /**
+     * This is an ugly hack, but currently we have no
+     * other way of accessing the underlying untyped
+     * context of an ActorContextAdapter.
+     */
+    private def toUntyped(ctx: at.ActorContext[A]): au.ActorContext = {
+      ActorContextAccess.getUntypedActorContext(ctx)
+    }
+  }
+}
+

--- a/src/main/scala/com/nokia/ntp/ct/persistence/persistentBehavior.scala
+++ b/src/main/scala/com/nokia/ntp/ct/persistence/persistentBehavior.scala
@@ -37,11 +37,9 @@ trait SelfDeployingBehavior[A] extends at.Behavior[A] {
   def props: au.Props @@ A =
     cachedProps
 
-  def deployInto(system: au.ActorSystem, name: String = ""): at.ActorRef[A] = {
-    at.adapter.actorRefAdapter(
-      if (name.nonEmpty) system.actorOf(props, name)
-      else system.actorOf(props)
-    )
+  def deployInto[D](deployer: D, name: String = "")(implicit D: ActorDeployer[D]): at.ActorRef[A] = {
+    if (name.nonEmpty) D.deploy(deployer)(props, name)
+    else D.deployAnonymous(deployer)(props)
   }
 
   protected def untypedProps: au.Props

--- a/src/test/scala/com/nokia/ntp/ct/persistence/DeploymentSpec.scala
+++ b/src/test/scala/com/nokia/ntp/ct/persistence/DeploymentSpec.scala
@@ -1,0 +1,114 @@
+/*
+ * Copyright 2016 Nokia Solutions and Networks Oy
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.nokia.ntp.ct
+package persistence
+
+import akka.{ typed => at }
+import at.AskPattern._
+
+class DeploymentSpec extends AbstractPersistenceSpec {
+
+  import DeploymentSpec._
+
+  "SelfDeployingBehavior" should "be able to start in an (untyped) ActorSystem" in {
+    val name = "test1"
+    val ref = child(name).deployInto(this.system, name)
+    ref.?[Int](Num(42, _)).futureValue should === (43)
+    ref.?[Int](Num(42, _)).futureValue should === (44)
+    ref ! Stop
+    awaitTermination(ref).futureValue
+
+    val ref2 = child(name).deployInto(this.system, name)
+    ref2.?[Int](Num(5, _)).futureValue should === (8)
+    ref2.?[Int](Num(5, _)).futureValue should === (9)
+    ref2 ! Stop
+    awaitTermination(ref2).futureValue
+  }
+
+  it should "be able to start in an ActorContext (i.e., as a child actor)" in {
+    val name = "test2"
+    val ref = parent(name).deployInto(this.system, name)
+    ref.?[Int](Num(42, _)).futureValue should === (43)
+    ref.?[Int](Num(42, _)).futureValue should === (44)
+    ref ! Stop
+    awaitTermination(ref).futureValue
+
+    val ref2 = parent(name).deployInto(this.system, name)
+    ref2.?[Int](Num(9, _)).futureValue should === (12)
+    ref2.?[Int](Num(9, _)).futureValue should === (13)
+    ref2 ! Stop
+    awaitTermination(ref2).futureValue
+  }
+}
+
+object DeploymentSpec {
+
+  sealed trait Mesg
+  final case class Num(i: Int, replyTo: at.ActorRef[Int]) extends Mesg
+  final case object Stop extends Mesg
+
+  sealed trait ParentState
+  final case object Uninitialized extends ParentState
+  final case class Initialized(child: at.ActorRef[Mesg]) extends ParentState
+
+  def parent(pid: PersistenceId) = PersistentFull[Mesg, at.ActorRef[Mesg], ParentState](
+    Uninitialized,
+    _ => s"${pid}-parent",
+    recovery = Recovery(
+      (_, msg, ctx) => {
+        val childRef = child(pid).deployInto(ctx)
+        Initialized(childRef)
+      }
+    )
+  ) { state => ctx =>
+      PersistentBehavior.Total {
+        case Stop =>
+          ctx.stop
+        case n @ Num(_, _) =>
+          state match {
+            case Uninitialized =>
+              val childRef = child(pid).deployInto(ctx.ctx)
+              childRef ! n
+              ctx.apply(childRef, sync = true)
+            case Initialized(child) =>
+              child ! n
+              ctx.same
+          }
+      }
+    }
+
+  def child(pid: PersistenceId) = Persistent[Mesg, Unit, Int](
+    0,
+    _ => s"${pid}-child"
+  ) { state => ctx =>
+      PersistentBehavior.Total {
+        case Stop =>
+          ctx.stop
+        case Num(i, r) =>
+          ctx.apply(()).map { state =>
+            r ! i + state
+            state
+          }
+      }
+    }
+
+  implicit val childUpdater: Update[Int, Unit] =
+    Update.instance((s, _) => s + 1)
+
+  implicit val parentUpdater: Update[ParentState, at.ActorRef[Mesg]] =
+    Update.instance((_, e) => Initialized(e))
+}

